### ISSUE TITLE
New Rule: no-destructuring-in-function-parameters

### ISSUE THREAD

**Issue by dannybloe** [eslint/eslint#15838]
### Rule details

It should prevent/control the use of descructuring arrays/objects in parameter definition of functions

### Related ECMAScript feature

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#unpacking_properties_from_objects_passed_as_a_function_parameter

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

```js
// should warn or error for unpacking the parameter in the function definition
function zork({ param1, param2 }) {
    console.log(param1, param2);
}

zork({ param1: 'a', param2: 'b' });
```

### Why should this rule be in the core instead of a plugin?

Many consider it a bad practice to use this pattern because it introduces many problems and inconveniences like:

 1. Debugging the function zork in the example is harder because you cannot see what parameter is actually passed when setting a breakpoint in the function body. 
 2. Typing is a lot harder when using TypeScript. 
 3. Many devs want a function to define exactly what it requires and be strict about it. Zork above can accept any object as long as it has param1 and param2 in there. 
 4. Many IDEs like IntelIJ show parameter names in function calls as prefix to values. That doesn't work for functions like zork. 

All in all, to me and others this is a horrible pattern when defining functions and it would be nice if eslint could enforce or at least warn you about this.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

### Additional comments

The suggested rule-name is probably not the best way to name this rule 🙂

**Comment by aladdin-add** [comment]
Per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies), we can only accept new rules if they relate to new ECMAScript features that reached stage 4 in the preceding 12 months, so this proposal doesn't qualify. You can always make a custom rule and publish a plugin.
